### PR TITLE
[POC] use Remark

### PR DESCRIPTION
**Pull request by DeltaEvo**
Hello @bryphe 

I think that to make a language server for markdown is a very great idea

But i saw that you started to implement a parser from scratch with regex, I thinked that with the awesome markdown parsers already available we could use one.

I started some search and I found [remark](https://github.com/remarkjs/remark) a very great project that use [unified](https://github.com/unifiedjs/unified) with [mdast](https://github.com/syntax-tree/mdast), these two projects provide standard interface to work with text parsing and markdown AST (so we can use great libraries like [unist-util-visit](https://github.com/syntax-tree/unist-util-visit)). On top of that remark's team has build lints [remark-lint](https://github.com/remarkjs/remark-lint) so i used them to provide validation and they have a great stringifier with a lot of option [remark-stringify](https://github.com/remarkjs/remark/tree/master/packages/remark-stringify#api)

It's still WIP because I need to implement:
-  [x] MarkdownDocument cache
-  [x] Try on large markdown document, and determine if we need incremental parsing (with standard MDAst this could be possible to implement it and to continue to use remark-lint and remark-stringify)
-  [ ] Range formatting (Find the nodes between the range in the AST)
-  [x] Unit Test

PS: It's my first real use of typescript so correct me if I made some errors

### CODE REVIEW

**Comment by bryphe** [comment]
Hi @DeltaEvo ,

Awesome work! Thank you for contributing!

> But i saw that you started to implement a parser from scratch with regex, I thinked that with the awesome markdown parsers already available we could use one.

Agree 110%. I would prefer not to own the parsing if possible. I didn't know about `unified` and `remark-parse` until I saw your PR. Using a real parser and getting the AST is a much better approach than the regexes I had, and will enable us to do more interesting things with the language server. Very cool!

> On top of that remark's team has build lints remark-lint so i used them to provide validation and they have a great stringifier with a lot of option remark-stringify

Cool, definitely makes sense to leverage these!

> PS: It's my first real use of typescript so correct me if I made some errors

Code looks great to me - I didn't see any problems. Looks like you picked it up fast 😄

**Comment by bryphe** [comment]
By the way @DeltaEvo , I hope you don't mind - I sent you a 'Collaborator' invite. You know more than I do about language servers... so doesn't make sense for you to wait on me for PRs 😄

**Comment by DeltaEvo** [comment]
Thanks @bryphe,
I accepted the invitation, and I think that a review from another developer is always valuable so I will wait for a final review because even if I have some knowledge in language server my code isn't perfect

**Comment by bryphe** [comment]
Sounds great @DeltaEvo - I'll do the same as well with my PRs here.

File: src/typing.d.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    freeze(): Processor
+  }
+  type ProcessorFunc = () => Processor
+  
+  const unified: Processor
+
+  export default unified
+}
+
+declare module 'vfile' {
```
Comment (bryphe) [inline src/typing.d.ts]: FYI, It looks like for the `vfile` module, there is actually a typings available - you can `npm install @types/vfile` and then you won't need to include the `declare module 'vfile'` block.

Comment (DeltaEvo) [reply src/typing.d.ts]: Thanks

**Comment by DeltaEvo** [comment]
Hello @bryphe 
I'm sorry to have been so long
I don't have time to implement range formatting for the moment so I just disabled it
However I think this PR is Ready ;)

**Comment by bryphe** [comment]
Hi @DeltaEvo ,

> I'm sorry to have been so long

Oh no worries, thank you for taking this on!

> I don't have time to implement range formatting for the moment so I just disabled it
However I think this PR is Ready ;)

Sounds great - looks good to me 👍 Will be a solid foundation for us to build on!

**Pull request merged**